Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -568,6 +509,64 @@ fileprivate extension Product {
     }
 }
 
+private class DuplicateProductsChecker {
+    var packageIDToBuilder = [String: ResolvedPackageBuilder]()
+    var checkedPkgIDs = [String]()
+
+    init(packageBuilders: [ResolvedPackageBuilder]) {
+        for packageBuilder in packageBuilders {
+            let pkgID = packageBuilder.package.identity.description.lowercased()
+            packageIDToBuilder[pkgID] = packageBuilder
+        }
+    }
+
+    func run(lookupByProductIDs: Bool = false, observabilityScope: ObservabilityScope) throws {
+        var productToPkgMap = [String: [String]]()
+        for (_, pkgBuilder) in packageIDToBuilder {
+            let useProductIDs = pkgBuilder.package.manifest.toolsVersion >= .vNext || lookupByProductIDs
+            let depProductRefs = pkgBuilder.package.targets.map{$0.dependencies}.flatMap{$0}.compactMap{$0.product}
+            for depRef in depProductRefs {
+                if let depPkg = depRef.package?.lowercased() {
+                    checkedPkgIDs.append(depPkg)
+                    let depProductIDs = packageIDToBuilder[depPkg]?.package.products.filter { $0.ID == depRef.ID }.map { useProductIDs && $0.isDefaultLibrary ? $0.ID : $0.name } ?? []
+                    for depID in depProductIDs {
+                        productToPkgMap[depID, default: []].append(depPkg)
+                    }
+                } else {
+                    let depPkgs = pkgBuilder.dependencies.filter{$0.products.contains{$0.product.name == depRef.name}}.map{$0.package.identity.description.lowercased()}
+                    productToPkgMap[depRef.name, default: []].append(contentsOf: depPkgs)
+                    checkedPkgIDs.append(contentsOf: depPkgs)
+                }
+            }
+
+            for (depIDOrName, depPkgs) in productToPkgMap.filter({Set($0.value).count > 1}) {
+                if useProductIDs {
+                    throw PackageGraphError.duplicateProduct(product: depIDOrName, packages: depPkgs.sorted())
```

Comment:
We might want to have the error be consistent regardless of package version? E.g. by using separate package identity and product name arguments.

Reply:
Btw, should we throw an error instead of emitting the error regardless of the package version? If the error is emitted as in old versions, it can't really resolve the graph anyway and by simply throwing the error we could also remove the logic that tries to remove duplicates after the error emitting.